include/lo2s/perf/reader.hpp

```diff
@@ -0,0 +1,368 @@
+/*
+ * This file is part of the lo2s software.
+ * Linux OTF2 sampling
+ *
+ * Copyright (c) 2017,
+ *    Technische Universitaet Dresden, Germany
+ *
+ * lo2s is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * lo2s is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with lo2s.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#pragma once
+
+#include <cstdint>
+#include <set>
+#include <variant>
+
+#include <lo2s/config.hpp>
+#include <lo2s/error.hpp>
+#include <lo2s/execution_scope.hpp>
+#include <lo2s/perf/tracepoint/format.hpp>
+#include <lo2s/perf/util.hpp>
+
+#ifndef USE_HW_BREAKPOINT_COMPAT
+extern "C"
+{
+#include <linux/hw_breakpoint.h>
+}
+#else
+extern "C"
+{
+#include <sys/types.h>
+#include <sys/wait.h>
+}
+#endif
+
+namespace lo2s
+{
+namespace perf
+{
+
+enum class Availability
+{
+    UNAVAILABLE,
+    SYSTEM_MODE,
+    PROCESS_MODE,
+    UNIVERSAL
+};
+
+class PerfEventInstance;
+
+/**
+ * Base class for all Event types
+ * contains common attributes
+ */
+class PerfEvent
+{
+public:
+    PerfEvent(bool enable_on_exec);
+    PerfEvent();
+
+    /**
+     * returns an opened instance of any PerfEvent object
+     * @param group_fd file descriptor of the leader event, omit if there is no leader
+     */
+    PerfEventInstance open(std::variant<Cpu, Thread> location, int group_fd = -1,
+                           int cgroup_fd = config().cgroup_fd);
+
+    perf_event_attr& get_attr()
+    {
+        return attr_;
+    }
+
+    double get_scale() const
+    {
+        return scale_;
+    }
+
+    void set_scale(double scale)
+    {
+        scale_ = scale;
+    }
+
+    std::string get_unit() const
+    {
+        return unit_;
+    }
+
+    void set_unit(std::string unit)
+    {
+        unit_ = unit;
+    }
+
+    friend bool operator==(const PerfEvent& lhs, const PerfEvent& rhs)
+    {
+        return !memcmp(&lhs.attr_, &rhs.attr_, sizeof(struct perf_event_attr));
+    }
+
+    friend bool operator<(const PerfEvent& lhs, const PerfEvent& rhs)
+    {
+        return memcmp(&lhs.attr_, &rhs.attr_, sizeof(struct perf_event_attr));
+    }
+
+    ~PerfEvent()
+    {
+    }
+
+protected:
+    struct perf_event_attr attr_;
+    double scale_ = 1;
+    std::string unit_ = "#";
+};
+
+/**
+ * Contains predefined events, like the ones listed by --list-events
+ * @note define the type by either calling as_counter() or as_group_leader() after creating an
+ * object
+ */
+class PredefEvent : public PerfEvent
```

Comment:
Since a PredefEvent has as_counter() and as_group_leader() this would also be moved into PerfEvent and it feels weird that you could then make a TimeEvent and call as_counter() on it, for example.

Reply:
`PredefEvent` is not needed anymore. Should `PerfEvent` still be renamed to `Event`?

Reply:
I think so. This is already in the perf namespace so it being a perf::PerfEvent instead of a perf::Event seems a bit redundant.